cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,336 @@
+                tree.add_node(unique_id)
+                tree.nodes[unique_id]["birth_scale"] = (
+                    lineage["birth_scale"] * total_birth_mutation_strength
+                )
+                tree.add_edge(
+                    lineage["id"], unique_id, weight=birth_waiting_time
+                )
+                tree.nodes[unique_id]["total_time"] = (
+                    birth_waiting_time + lineage["total_time"]
+                )
+                # Add the newly generated cell to the list of living lineages
+                current_lineages.append(
+                    {
+                        "id": unique_id,
+                        "birth_scale": lineage["birth_scale"]
+                        * total_birth_mutation_strength,
+                        "total_time": birth_waiting_time
+                        + lineage["total_time"],
+                    }
+                )
+                return unique_id + 1
+            else:
+                return unique_id
+
+        # Instantiate the implicit root
+        tree = nx.DiGraph()
+        tree.add_node(0)
+        tree.nodes[0]["birth_scale"] = birth_scale_param
+        tree.nodes[0]["total_time"] = 0
+        current_lineages = []
+        starting_lineage = {
+            "id": 0,
+            "birth_scale": birth_scale_param,
+            "total_time": 0,
+        }
+
+        # Sample the waiting time until the first division
+        unique_id = 1
+        unique_id = sample_event(unique_id, starting_lineage)
+
+        # Perform the process until there are no active extant lineages left
+        while len(current_lineages) > 0:
+            # If number of extant lineages is the stopping criterion, at the
+            # first instance of having n extant tips, stop the experiment
+            # and set the total lineage time for each lineage to be equal to
+            # the minimum, to produce ultrametric trees. Also, the birth_scale
+            # parameter of each leaf is rolled back to equal its parent's.
+            if num_extant:
+                if len(current_lineages) == num_extant:
+                    min_total_time = min(
+                        [i["total_time"] for i in current_lineages]
+                    )
+                    for remaining_lineage in current_lineages:
+                        parent = list(
+                            tree.predecessors(remaining_lineage["id"])
+                        )[0]
+                        tree.edges[parent, remaining_lineage["id"]][
+                            "weight"
+                        ] += (min_total_time - remaining_lineage["total_time"])
+                        tree.nodes[remaining_lineage["id"]][
+                            "birth_scale"
+                        ] = tree.nodes[parent]["birth_scale"]
+                    break
+            # If extant tips are the stopping criteria, pop the minimum age
+            # lineage at each step
```

Comment:
Popping the minimum age lineage works regardless of whether the stopping condition is num_extant or experiment_time.